win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+                    {
+                        /*Codes_SRS_FILE_43_036: [ If the call to read the file fails, file_read_async shall fail and return FILE_READ_ASYNC_READ_ERROR. ]*/
+                        /*Codes_SRS_FILE_WIN32_43_031: [ If ReadFile fails synchronously and GetLastError does not indicate ERROR_IO_PENDING then file_read_async shall fail, call CancelThreadpoolIo and return FILE_READ_ASYNC_WRITE_ERROR. ]*/
+                        LogError("Failure in ReadFile");
+                        CancelThreadpoolIo(handle->ptp_io);
+                        result = FILE_READ_ASYNC_READ_ERROR;
+                    }
+                }
+                else
+                {
+                    /*Codes_SRS_FILE_43_016: [ file_read_async shall call user_callback passing user_context and success depending on the success of the asynchronous read operation.]*/
+                    /*Codes_SRS_FILE_43_031: [ file_read_async shall succeed and return FILE_READ_ASYNC_OK. ]*/
+                    /*Codes_SRS_FILE_WIN32_43_032: [ If ReadFile succeeds synchronously then file_read_async shall succeed, call CancelThreadpoolIo, call user_callback and return FILE_READ_ASYNC_OK. ]*/
+                    CancelThreadpoolIo(handle->ptp_io);
+                    user_callback(user_context, true);
+                    result = FILE_READ_ASYNC_OK;
+                }
+                if (!callback_will_be_called)
+                {
+                    if (!CloseHandle(io->ov.hEvent))
+                    {
+                        LogError("Failure in CloseHandle");
+                    }
+                }
+            }
+            if (!callback_will_be_called)
+            {
+                free(io);
+            }
+        }
+    }
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int, file_extend, FILE_HANDLE, handle, uint64_t, desired_size)
+{
+    (void)handle;
+    (void)desired_size;
+    /*Codes_SRS_FILE_WIN32_43_050: [ file_extend shall return 0. ]*/
+    return 0;
+}
+
+static void on_file_io_complete_win32(
+    PTP_CALLBACK_INSTANCE instance,
+    PVOID context,
+    PVOID overlapped,
+    ULONG io_result,
+    ULONG_PTR number_of_bytes_transferred,
+    PTP_IO io
+)
+{
+    (void)instance;
+    (void)context;
+    (void)io_result;
+    (void)number_of_bytes_transferred;
+    (void)io;
+
+    void (*callback_function)();
+    void* callback_context;
+    HANDLE handle = NULL;
+    bool report_fault = false;
+
+    /*Codes_SRS_FILE_WIN32_43_034: [ on_file_io_complete_win32 shall recover the file handle, the number of bytes requested by the user, user_callback and user_context from the context containing overlapped. ]*/
+    FILE_WIN32_IO* io_context = CONTAINING_RECORD(overlapped, FILE_WIN32_IO, ov);
+    if (io_context->user_callback == NULL)
+    {
+        LogError("Could not recover user-specified callback");
+        report_fault = true;
+        callback_function = (FILE_REPORT_FAULT)(io_context->handle->user_report_fault_callback);

```

Comment:
>(FILE_REPORT_FAULT) [](start = 28, length = 19)

:/ casting a function pointer can only lead to bugs... I don't think we should use one variable to hold two different callbacks